sklearn/tree/_tree.pyx

```diff
@@ -255,7 +274,36 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
                 # inspection and interpretation
                 splitter.node_value(tree.value + node_id * tree.value_stride)
 
+                middle_value = tree._get_middle_value(node_id)
```

Comment:
This only needs to be called when `with_monotonic_cst` is True and `splitter.monotonic_cst[split.feature] == 0`.

Reply:
Did you mean that it only needs to be called when all the following conditions are met:
 - `is_leaf is False`
 - `with_monotonic_cst is True`
 - `splitter.monotonic_cst[split.feature] != 0`
 
?

Reply:
let's nest the `middle_value = tree._get_middle_value(node_id)` call under the conditions, only the branch where we need it. It's additional duplication but the code will be more efficient and readable.